h/schemas/oauth.py

```diff
@@ -55,23 +44,10 @@ class Schema(JSONSchema):
 
         validated_data = Schema().validate(data)
 
-        if validated_data["state"] != self._request.session.pop(
-            self.state_session_key, None
-        ):
+        if validated_data["state"] != expected_state:
```

Comment:
It's still up to the schema to validate the `state` param including that it matches the expected value that was retrieved from the session.

Part of the reason for this is that the schema raises a particular type of exception (`ValidationError`) that has special exception view handling, and I didn't want the view to have to raise its own `ValidationError` just for the case of `state` not matching the expected state.

Another reason is just that `OAuth2RedirectSchema.validate(params)` is supposed to return the validated params, so they should be fully valid, not just partially validated and requiring further validation by the view itself.

So I added a required `expected_state` arg: `OAuth2RedirectSchema.validate(params, expected_state)`.